### ISSUE TITLE
SA1300 misfires on members that implement interfaces

### ISSUE THREAD

**Issue by AArnott** [DotNetAnalyzers/StyleCopAnalyzers#771]
All rules that enforce naming conventions of members should *not* fire when the member provides an interface implementation. 

The rule *can* and probably should fire on the member in its interface declaration, but not on its implementations. For instance:

On the interface, the rule may fire:

```csharp
interface IVsCfg
{
    int get_DisplayName(out string name);
}
```

But don't blame the implementation of such a method:

```csharp
class MyVsCfg : IVsCfg
{
    public int get_DisplayName(out string name)
    {
        name = null;
        return 0;
    }
}
```

Another example of a place to avoid reporting a diagnostic is explicit interface implementations:

```csharp
class MyVsCfg : IVsCfg
{
    int IVsCfg.get_DisplayName(out string name)
    {
        name = null;
        return 0;
    }
}
```

**Comment by sharwell** [comment]
I was thinking about this yesterday actually. I was thinking we could do this:

1. Do not report a warning if the declaration uses `override`.
2. Do not report a warning if the declaration *explicitly* implements an interface member.

However, I'm not sure we should make an exception for declarations that implicitly implement an interface member.

**Comment by AArnott** [comment]
Adding `override` is a good idea.

> I'm not sure we should make an exception for declarations that implicitly implement an interface member.

Why not? The compelling case for me happens frequently: the implementing class actually implements *many* interfaces (e.g. `IVsCfg`, `IVsCfg2`, `IVsCfg3`) and they all derive from their predecessor. So using explicit interface implementation syntax means I have to define an extra method for every interface rather than just making the method `public` and let the compiler implement all interfaces at once.

"But it's worth it," I can hear folks say "...because if you make it public that means the bad name is also visible to folks just using your class." Ya, and I would go for that too. Except the class itself is `internal` so it's a non-issue. It is only there to implement the interfaces and not to be exposed itself.

**Comment by pdelvo** [comment]
I have a look at this

**Comment by AArnott** [comment]
One other reason I hit today---sometimes it's simply not an option at all to make it an explicit interface implementation. When an object is exposed to COM as an IDispatch object, the members must actually be public (not explicit interface implementation) or else IDispatch fails.

**Comment by AArnott** [comment]
Thanks!
